src/fromager/resolver.py

```diff
@@ -145,14 +146,27 @@ def identify(self, requirement_or_candidate: Requirement | Candidate) -> str:
     def get_extras_for(
         self,
         requirement_or_candidate: Requirement | Candidate,
-    ) -> tuple[str]:
+    ) -> tuple[str, ...]:
         # Extras is a set, which is not hashable
-        return tuple(sorted(requirement_or_candidate.extras))
+        if requirement_or_candidate.extras:
+            return tuple(sorted(list(requirement_or_candidate.extras)))
+        return tuple()
 
     def get_base_requirement(self, candidate: Candidate) -> Requirement:
         return Requirement(f"{candidate.name}=={candidate.version}")
 
-    def get_preference(self, identifier, resolutions, candidates, information, **kwds):
+    def get_preference(
+        self,
+        identifier: str,
+        resolutions: typing.Mapping[str, Candidate],
+        candidates: typing.Mapping[str, typing.Iterator[Candidate]],
```

Comment:
Why iterator? Could we use the CandidateMap type alias here?
```suggestion
        candidates: CandidateMap,
```

Reply:
I had just copied these types from resolvelib